JuliaLowering/test/scopes.jl

```diff
@@ -58,23 +58,292 @@ begin
 end
 """) === :outer_y
 
-# wrap expression in scope block of `scope_type`
-function wrapscope(ex, scope_type)
-    g = JuliaLowering.ensure_attributes(ex._graph, scope_type=Symbol)
-    ex = JuliaLowering.reparent(g, ex)
-    makenode(g, ex, K"scope_block", [ex._id], [:scope_type=>scope_type])
+#=
+| old\new: || global     | local | arg                | sparam             |
+|----------++------------+-------+--------------------+--------------------|
+| global   || no-op      | (*)   |                    |                    |
+| local    || error (*)  | no-op |                    |                    |
+| arg      || shadow(??) | error | error (not unique) |                    |
+| sparam   || shadow(??) | error | error (sparam/arg) | error (not unique) |
+=#
+@testset "Conflicts in the same local scope" begin
+
+    # no-op cases.  It would probably be clearer (but breaking) if these were
+    # errors like the conflict cases (two of the same decl should never do
+    # anything, and the user might be expecting two variables).
+    @testset "global,global" begin
+        s = "function (); global g; global g; 1; end"
+        @test JuliaLowering.include_string(test_mod, s) isa Function
+    end
+    @testset "local,local" begin
+        s = "function (); local l; local l; end"
+        @test JuliaLowering.include_string(test_mod, s) isa Function
+    end
+
+    # locals may not overlap args/sparams/globals
+    @testset "global,local/local,global" begin
+        s = "function (); global g; local g; end"
+        @test_throws LoweringError JuliaLowering.include_string(test_mod, s)
+    end
+    @testset "arg,local" begin
+        s = "function (x); local x; end"
+        @test_throws LoweringError JuliaLowering.include_string(test_mod, s)
+    end
+    @testset "sparam,local" begin
+        s = "function (a::s) where {s}; local s; end"
+        @test_throws LoweringError JuliaLowering.include_string(test_mod, s)
+    end
+
+    # globals may overlap args or sparams (buggy?)  TODO: decide whether it's
+    # worth replicating this behaviour.  We would likely need to copy the way
+    # flisp nests an extra scope block in every lambda.
```

Comment:
Eek.  I can't recall right now whether I intentionally tried to depart from the weird flisp behavior or not.  seems potentially breaking to be different (if logical) - may need to be changed via syntax evolution I guess.

Reply:
Sounds reasonable.